README.rst
pkgwat.api
==========

:Author: Ralph Bean <rbean@redhat.com>

.. comment: split here

Python API for `pkgwat <http://pypi.python.org/pypi/pkgwat.cli>`_.

Documentation:  http://pkgwat.rtfd.org


Setting up development environment
----------------------------------

Make sure you have ``virtualenv`` installed and create a new venv::

  $ virtualenv env
  $ source env/bin/activate
  $ pip install -e .


Running the test suite
----------------------

Make sure you have ``tox`` installed and run it (outside of any virtualenv)::

  $ tox


Cutting a new release
---------------------

In order to generate a list of changes to be placed into ``CHANGELOG.rst``, use
the following command (this example generates the log between ``0.12`` and
``HEAD`` refs)::

  git log --reverse --format=format:'- %s `%h <https://github.com/fedora-infra/pkgwat.api/commit/%H>`_' 0.12..HEAD


CHANGELOG.rst
Changelog
=========

0.13
----

- Merge branch 'release/0.12' into develop `0fad55f <https://github.com/fedora-infra/pkgwat.api/commit/0fad55fb785cf8192441a0b36e95b6b9823ca9d7>`_
- nuke strip_tags in history() `1ae25c4 <https://github.com/fedora-infra/pkgwat.api/commit/1ae25c40e17e453ecfbf1f0358f0858316b0e7a2>`_
- Remove these deprecated functions as per #22. `b0e39cd <https://github.com/fedora-infra/pkgwat.api/commit/b0e39cd5ec491f2984d3cd87d3352579f9d4d679>`_
- Merge pull request #24 from fedora-infra/feature/remove-relationships `9189cde <https://github.com/fedora-infra/pkgwat.api/commit/9189cde0b3b591cc38815e2c56a10ba17b88e737>`_
- Remove the arch parameter since it is not supported by fedora-packages `b861a08 <https://github.com/fedora-infra/pkgwat.api/commit/b861a08094264ff823d2ba566f6166be832ec883>`_
- Add package_name to the changelog filters as it is expected by fedora-packages `c894505 <https://github.com/fedora-infra/pkgwat.api/commit/c8945059887fd3aa24546d66dbec2bb84ed5eaf5>`_
- The last owner of guake builds was releng not pingou `5951e82 <https://github.com/fedora-infra/pkgwat.api/commit/5951e82e4549eab0e7ec38c7841ec88b0de64086>`_
- Add support for tox test `1f5df9e <https://github.com/fedora-infra/pkgwat.api/commit/1f5df9efab7c9748ebbeca7b103e51c5e9006b0a>`_
- Add .cico.pipeline `166a935 <https://github.com/fedora-infra/pkgwat.api/commit/166a935d69708701fe047a17d0daad2c732c9a8c>`_
- Pass HOME env to tox runner `3c43c15 <https://github.com/fedora-infra/pkgwat.api/commit/3c43c15c670cfe8c201be7b10b3ed8e343b54f1a>`_
- fix parser crash `0807919 <https://github.com/fedora-infra/pkgwat.api/commit/0807919401911e68e1a872e1325e952d74e597ac>`_
- tests: update guake homepage `1235015 <https://github.com/fedora-infra/pkgwat.api/commit/12350153eff7dfd6a508309ec160b66d42439525>`_
- fix tox to run tests `4d3f233 <https://github.com/fedora-infra/pkgwat.api/commit/4d3f233af8d89995fd376985a2d2716a1ae334be>`_
- readme: add instructions to set up devenv and run tests `87d23e4 <https://github.com/fedora-infra/pkgwat.api/commit/87d23e48b54c9c5f8fbc492a77a5f004bb8da667>`_
- changelog: fix last releases merged together `8e7f10e <https://github.com/fedora-infra/pkgwat.api/commit/8e7f10efcd2e84c1e04eeb3d7bf154e19e7ed6b6>`_

0.12
----

- Temporarily remove dep on kitchen until py3 support can be sorted out. `0f4d864c4 <https://github.com/fedora-infra/pkgwat.api/commit/0f4d864c474280143d437982c8e97b0adc6f25d1>`_
- Guake actually only provides 2 things I guess. `cf0ce50a8 <https://github.com/fedora-infra/pkgwat.api/commit/cf0ce50a801ee0d58b53f45b8a1910dd17e40d75>`_
- Merge pull request #19 from fedora-infra/feature/nix-kitchen `d9be9c268 <https://github.com/fedora-infra/pkgwat.api/commit/d9be9c268b1de21874db6b49941e83f42f4a1662>`_

0.11
----

- ascii codec cant encode character `f4235e06e <https://github.com/fedora-infra/pkgwat.api/commit/f4235e06e66331d6828bfcd3b33c584c29c7294e>`_
- Merge pull request #17 from arielb2/bug-utf `8acb9bb21 <https://github.com/fedora-infra/pkgwat.api/commit/8acb9bb21688cbe585ba8e2d6ee36a73c5865c48>`_
- Add kitchen to the requires list.  From #17. `d0fffd8ae <https://github.com/fedora-infra/pkgwat.api/commit/d0fffd8ae562444b31f31c06c4ca296c1435a807>`_
- PEP8/cosmetic. `851424586 <https://github.com/fedora-infra/pkgwat.api/commit/851424586e2821fbda7fea05a736f3785447bfe5>`_
- Merge pull request #18 from fedora-infra/feature/cleanup `18c25c073 <https://github.com/fedora-infra/pkgwat.api/commit/18c25c073f347b78516ed22ea2ab2a2ddbf2b9dc>`_

0.10
----

- Add failing tests for tag stripping. `0ab263ba5 <https://github.com/fedora-infra/pkgwat.api/commit/0ab263ba527c9d46e8541278752cf6812693d169>`_
- Fix tag stripping once and for all. `ae12f6c28 <https://github.com/fedora-infra/pkgwat.api/commit/ae12f6c28acbc649fdde51a72bb9707800416fb1>`_
- Add one more test of strip_tags, just for good measure. `d64cb2cf6 <https://github.com/fedora-infra/pkgwat.api/commit/d64cb2cf67fda2a20ae0f53b697bbcae3540c5d1>`_
- Merge pull request #13 from fedora-infra/feature/fix-tag-stripping `d64da0a14 <https://github.com/fedora-infra/pkgwat.api/commit/d64da0a149610425782d7d907b45960620ad543d>`_
- unicode escape fix https://github.com/fedora-infra/fedora-packages/issues/28 `b058ce006 <https://github.com/fedora-infra/pkgwat.api/commit/b058ce006e1ea760088154efe9f6f589c2cf2d2a>`_
- Handle erroneous html_parser exception on py2.6. `f0057a18c <https://github.com/fedora-infra/pkgwat.api/commit/f0057a18ca8118eaa183ec6970a3a325f1962bdf>`_
- Merge pull request #16 from fedora-infra/feature/py2.6-html_parser-fix `e979a0293 <https://github.com/fedora-infra/pkgwat.api/commit/e979a029308cc9a8a94def7333d754d5c5887c5f>`_
- Remove unnecessary markup. `581b7dd4e <https://github.com/fedora-infra/pkgwat.api/commit/581b7dd4e24801b8865e7239d70a644808a14833>`_
- Quotes. `ec25d0de1 <https://github.com/fedora-infra/pkgwat.api/commit/ec25d0de171ab426a0232281dcb763c80fd7425d>`_

0.9
---

- Allow .get to return subpackages. `1db230381 <https://github.com/fedora-infra/pkgwat.api/commit/1db23038157cf2513304bf56aac33ca74b039b77>`_
- Merge pull request #8 from fedora-infra/feature/get-subpackages `84e913fe1 <https://github.com/fedora-infra/pkgwat.api/commit/84e913fe1870a8748a1c2d52cbf790d5289076ac>`_
- Packages with different names `3d2930649 <https://github.com/fedora-infra/pkgwat.api/commit/3d293064933ef50a81a0ee6853fc101da40f1c09>`_
- Merge pull request #10 from arielb2/3d293064933ef50a81a0ee6853fc101da40f1c09 `f6ac43b46 <https://github.com/fedora-infra/pkgwat.api/commit/f6ac43b461ea8239bfa0c46e14045d01d7e7fc72>`_
- Add a history function to query datagrepper. `a3cd8950f <https://github.com/fedora-infra/pkgwat.api/commit/a3cd8950fd72c2e7bc33a61212858c257fc74856>`_
- Add a docstring for pkgwat.rtfd.org/ `30099ade4 <https://github.com/fedora-infra/pkgwat.api/commit/30099ade45068a8cf4bf20f8c9c99bcd9c389d55>`_

tests.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-

""" Basic unit-tests for the pkgwat API. """

import unittest
import six

from nose.tools import raises, eq_

from pkgwat.api import (
    get,
    bugs,
    builds,
    changelog,
    contents,
    releases,
    search,
    updates,
    dependencies,
    dependants,
    provides,
    obsoletes,
    conflicts,
    history,
)

import pkgwat.api.utils

PKG = 'guake'


class APItests(unittest.TestCase):
    """ Unit-tests for the pkgwat API. """

    def assert_keys_in_dict(self, d, expected):
        """ Utility for comparing dict keys. """
        self.assertTrue(all([key in d for key in expected]))
        self.assertTrue(all([key in expected for key in d]))

    def test_bugs(self):
        """ Test the bugs function from the API. """
        ## Let's assume there will always be at least one bug against
        ## guake.
        guake_bugs = bugs(PKG, release='all')
        self.assertTrue(len(guake_bugs) > 0)

    def test_builds(self):
        """ Test the builds function from the API. """
        guake_build = builds(PKG, state='all')
        self.assertEqual(guake_build['rows'][0]['package_name'], PKG)
        self.assertEqual(guake_build['rows'][0]['owner_name'], 'releng')
        self.assertEqual(guake_build['rows'][0]['package_id'], 6454)

    def test_builds_completed(self):
        """ Test the builds function from the API for the builds that
        completed.
        """
        guake_build = builds(PKG, state='1')
        self.assertEqual(guake_build['rows'][0]['state_str'], 'complete')
        self.assertEqual(guake_build['rows'][0]['package_id'], 6454)
        self.assertEqual(guake_build['rows'][0]['owner_name'], 'releng')

    def test_changelog(self):
        """ Test the changelog function from the API. """
        guake_changelog = changelog(PKG)
        self.assertTrue(len(guake_changelog['rows']) > 0)
        expected_keys = [six.u(key) for key in [
            'display_date',
            'changelog',
            'text',
            'author',
            'version',
            'date',
            'email',
        ]]
        self.assert_keys_in_dict(guake_changelog['rows'][0], expected_keys)

    def test_contents(self):
        """ Test the contents function from the API. """
        guake_contents = contents(PKG)
        expected_keys = [six.u('state'), six.u('data'), six.u('children')]
        self.assert_keys_in_dict(guake_contents[0], expected_keys)
        self.assertEqual(guake_contents[0]['state'], 'open')
        expected_data = {
            six.u('icon'): six.u('jstree-directory'),
            six.u('title'): six.u('usr')
        }
        self.assertEqual(guake_contents[0]['data'], expected_data)

    def test_releases(self):
        """ Test the releases function from the API. """
        guake_releases = releases(PKG)
        expected_keys = [six.u(key) for key in [
            'release', 'testing_version', 'stable_version'
        ]]
        self.assert_keys_in_dict(guake_releases['rows'][0], expected_keys)
        self.assertEqual(guake_releases['rows'][0]['release'], 'Rawhide')

    def test_search(self):
        """ Test the search function of the API. """
        guake_search = search(PKG)
        self.assertEqual(guake_search['rows'][0]['upstream_url'],
                         'http://guake-project.org/')
        self.assertEqual(guake_search['rows'][0]['devel_owner'],
                         'pingou')
        self.assertEqual(guake_search['rows'][0]['link'], PKG)

    def test_updates(self):
        """ Test the updates function of the API. """
        guake_updates = updates(PKG)
        expected_keys = [six.u(key) for key in [
            'status',
            'date_pushed',
            'name',
            'package_name',
            'title',
            'versions',
            'actions',
            'date_submitted_display',
            'karma_str',
            'dist_updates',
            'releases',
            'details',
            'request_id',
            'date_pushed_display',
            'karma_level',
            'id',
            'nvr',
        ]]

        self.assert_keys_in_dict(guake_updates['rows'][0], expected_keys)
        self.assertEqual(guake_updates['rows'][0]['name'], PKG)
        self.assertTrue(guake_updates['rows'][0]['title'].startswith(PKG))
        self.assertEqual(guake_updates['rows'][0]['package_name'], PKG)

    @raises(NotImplementedError)
    def test_dependencies(self):
        pkgwat_deps = dependencies("pkgwat")
        self.assertEqual(len(pkgwat_deps['rows']), 5)

    @raises(NotImplementedError)
    def test_dependants(self):
        pkgwat_dependants = dependants("python-pkgwat-api")
        self.assertEqual(len(pkgwat_dependants['rows']), 2)

    @raises(NotImplementedError)
    def test_provides(self):
        guake_provides = provides(PKG, version="0.4.2-6.fc17", arch="x86_64")
        self.assertEqual(len(guake_provides['rows']), 2)

    @raises(NotImplementedError)
    def test_obsoletes(self):
        guake_obsoletes = obsoletes(PKG, version="0.4.2-6.fc17", arch="x86_64")
        self.assertEqual(len(guake_obsoletes['rows']), 0)

    @raises(NotImplementedError)
    def test_conflicts(self):
        guake_conflicts = conflicts(PKG, version="0.4.2-6.fc17", arch="x86_64")
        self.assertEqual(len(guake_conflicts['rows']), 0)

    def test_get(self):
        """ Test the get function of the API. """
        guake_hit = get(PKG)
        self.assertEqual(guake_hit['upstream_url'],
                         'http://guake-project.org/')
        self.assertEqual(guake_hit['devel_owner'],
                         'pingou')
        self.assertEqual(guake_hit['link'], PKG)

    def test_history(self):
        """ Test the history function of the API. """
        results = history(PKG)
        assert len(results['raw_messages']) > 0

    @raises(KeyError)
    def test_get_fail(self):
        get("this-package-does-not-exist")

    def test_strip_together(self):
        string = six.u("Levio<em>sa</em>")
        target = six.u("Leviosa")
        actual = pkgwat.api.utils.strip_tags(string)
        eq_(target, actual)

    def test_strip_separate(self):
        string = six.u("I <em>told</em> you so!")
        target = six.u("I told you so!")
        actual = pkgwat.api.utils.strip_tags(string)
        eq_(target, actual)

    def test_strip_beginning(self):
        string = six.u('<a href="http:/..">fedora-tagger</a> is great')
        target = six.u('fedora-tagger is great')
        actual = pkgwat.api.utils.strip_tags(string)
        eq_(target, actual)


if __name__ == '__main__':
    SUITE = unittest.TestLoader().loadTestsFromTestCase(APItests)
    unittest.TextTestRunner(verbosity=2).run(SUITE)


pkgwat/api/utils.py
import six

from six.moves import html_parser


class MLStripper(html_parser.HTMLParser, object):
    strict = False

    def __init__(self):
        super(MLStripper, self).__init__()
        self.reset()
        self.fed = []

    def handle_data(self, d):
        self.fed.append(d)

    def get_data(self):
        return ''.join([item for item in self.fed])


def strip_tags(d):
    """ Recursively strip XML tags from a dict, list, or string. """

    if type(d) == dict:
        return dict(((k, strip_tags(v)) for k, v in d.items()))

    if type(d) == list:
        return [strip_tags(element) for element in d]

    if isinstance(d, six.text_type):
        s = MLStripper()
        try:
            s.feed(d)
        except html_parser.HTMLParseError:
            return d
        else:
            result = s.get_data()
            if isinstance(result, six.text_type):
                return result
            else:
                return result.decode('utf-8')

    return d


pkgwat/api/api.py

import collections
import json
import requests

import pkgwat.api
import pkgwat.api.utils

import requests.defaults
requests.defaults.defaults['base_headers']['User-Agent'] = \
        'pkgwat/' + pkgwat.api.__version__


# TODO -- Tie this into cliff's verbosity options
DEBUG_REQUESTS = False
if DEBUG_REQUESTS:
    import sys

    class myobj(object):
        def write(self, message):
            print("DEBUG:", message)

    requests.defaults.defaults['verbose'] = myobj()


BASE_URL = "https://apps.fedoraproject.org/packages/fcomm_connector"

koji_build_states = collections.OrderedDict((
    ('all', ''),
    ('building', '0'),
    ('success', '1'),
    ('failed', '3'),
    ('cancelled', '4'),
    ('deleted', '2'),
))

bodhi_releases = [
    "all",
    "f17",
    "f16",
    "f15",
    "el6",
    "el5",
]

bodhi_statuses = [
    "all",
    "stable",
    "testing",
    "pending",
    "obsolete",
]

yum_releases = [
    'Rawhide',
    'Fedora 16',
    'Fedora 16 Testing',
    'Fedora 15',
    'Fedora 15 Testing',
    'Fedora 14',
    'Fedora 14 Testing',
]

yum_arches = [
    'x86_64',
    'i686',
]

bugzilla_releases = collections.OrderedDict((
    ('all', ''),
    ('f17', '17'),
    ('f16', '16'),
    ('f15', '15'),
    ('el6', '6'),
    ('el5', '5'),
))


def _make_request(path, query, strip_tags):
    query_as_json = json.dumps(query)
    url = "/".join([BASE_URL, path, query_as_json])
    response = requests.get(url)
    d = json.loads(response.text)

    if strip_tags:
        d = pkgwat.api.utils.strip_tags(d)

    return d


def search(pattern, rows_per_page=10, start_row=0, strip_tags=True):
    path = "xapian/query/search_packages"
    query = {
        "filters": {"search": pattern},
        "rows_per_page": rows_per_page,
        "start_row": start_row,
    }

    return _make_request(path, query, strip_tags)


def releases(package, rows_per_page=10, start_row=0, strip_tags=True):
    path = "bodhi/query/query_active_releases"
    query = {
        "filters": {"package": package},
        "rows_per_page": rows_per_page,
        "start_row": start_row,
    }

    return _make_request(path, query, strip_tags)


def builds(package, state='all', rows_per_page=10,
             start_row=0, strip_tags=True):

    if state not in koji_build_states.values():
        state = koji_build_states[state]

    path = "koji/query/query_builds"
    query = {
        "filters": {
            "package": package,
            "state": state,
        },
        "rows_per_page": rows_per_page,
        "start_row": start_row,
    }

    return _make_request(path, query, strip_tags)


def updates(package, release="all", status="all", rows_per_page=10,
             start_row=0, strip_tags=True):

    if release not in bodhi_releases:
        raise ValueError("Invalid bodhi release. %r %r" % (
            release, bodhi_releases))

    if status not in bodhi_statuses:
        raise ValueError("Invalid bodhi status. %r %r" % (
            status, bodhi_statuses))

    if release == "all":
        release = ""

    if status == "all":
        status = ""

    path = "bodhi/query/query_updates"
    query = {
        "filters": {
            "package": package,
            "release": release,
            "status": status,
        },
        "rows_per_page": rows_per_page,
        "start_row": start_row,
    }

    return _make_request(path, query, strip_tags)


def bugs(package, release="all", rows_per_page=10,
         start_row=0, strip_tags=True):

    if release not in bugzilla_releases.values():
        release = bugzilla_releases[release]

    path = "bugzilla/query/query_bugs"
    query = {
        "filters": {
            "package": package,
            "version": release,
        },
        "rows_per_page": rows_per_page,
        "start_row": start_row,
    }

    return _make_request(path, query, strip_tags)


def contents(package, arch="x86_64", release="Rawhide", strip_tags=True):
    # This one behaves a little differently

    if release not in yum_releases:
        raise ValueError("Invalid yum release. %r %r" % (
            release, yum_releases))

    if arch not in yum_arches:
        raise ValueError("Invalid yum arch.  %r %r" % (
            arch, yum_arches))

    path = "yum/get_file_tree"
    query = {
        "package": package,
        "arch": arch,
        "repo": release,
    }
    url = "/".join([BASE_URL, path])
    response = requests.get(url, params=query)
    d = response.json

    if strip_tags:
        d = pkgwat.api.utils.strip_tags(d)

    return d


def changelog(package, rows_per_page=10, start_row=0, strip_tags=True):
    build_id = builds(package)['rows'][0]['build_id']

    path = "koji/query/query_changelogs"
    query = {
        "filters": {
            "build_id": build_id,
        },
        "rows_per_page": rows_per_page,
        "start_row": start_row,
    }

    return _make_request(path, query, strip_tags)
